com.ibm.wala.core/src/com/ibm/wala/ipa/callgraph/pruned/PrunedCallGraph.java

```diff
@@ -66,6 +67,12 @@ public void removeNodeAndEdges(CGNode n) throws UnsupportedOperationException {
 		return col.iterator();
 	}
 
+	@Override
+	public Stream<CGNode> stream() {
+		assert keep.stream().allMatch(cg::containsNode);
```

Comment:
While we're editing this file, can we patch up `iterator()` to do something like `Collections.unmodifiableSet(keep).iterator()`?  I don't see a good reason for it to do what it's doing now.  Unless iterating over `keep` is *really* slow?  But iterating over the whole call graph is probably slower.

Reply:
@msridhar, your suggested replacement is only equivalent if `keep` is a subset of the nodes in `cg`. However, I think that is not a safe assumption. The existing code strongly implies that it is OK for `keep` to contain nodes not in `cg`. For example, this must be why `PrunedCallGraph::containsNode()` checks both `cg` and `keep` instead of just checking `keep`.

I wrote the assertion `keep.stream().allMatch(cg::containsNode)` in `PrunedCallGraph::stream()`, but I now believe that this assertion is wrong in general. Rather, `PrunedCallGraph::stream()` should allow `keep` to include nodes not in `cg`, and should do the right thing in that case. The change we really need here is to make `PrunedCallGraph::stream()` look more like `PrunedCallGraph::iterator()`. That is, each of these methods should yield only nodes in the _intersection_ of `keep` and `cg`.

That being said, we can certainly avoid creating the temporary `col` container that holds this intersection. Instead, we just iterate through one of the two sets of nodes and only yield each item if it is also in the other set, thereby computing the intersection on the fly without ever reifying it.

@msridhar: your thoughts?

Reply:
Then again, `PrunedCallGraph::getNumberOfNodes()` is clearly assuming that `keep` is a subset of `cg`, since it returns the size of the former rather than the size of the intersection. OK, at this point I really do not know whether the subset assumption is valid or invalid. Some of the code hints one way, some hints the other. Help!

Reply:
We should have `keep` be a subset of `cg`.  I would say any code in contradiction of that assumption is buggy, and that we should add runtime checks (not just assertions) to fail fast if `keep` contains a node not in the `cg`, assuming it's not too expensive.

But, if desired we can address the broader problem in a follow-up issue / PR.

Reply:
Your proposed invariant sounds fine to me. But yes, I like the idea of dealing with that invariant separately. The logic in this file was already inconsistent as to the `keep` ⊆ `cg` question even before #427. We came across it due to examining #427 closely, but fixing it is a distinct matter and can sensibly be addressed separately.